### PR TITLE
OnPowerUse and AfterAttackRoll events

### PR DESCRIPTION
**Pull request by Gtaray**
Added two new events:
* OnPowerUse, which fires when a power on the PC actions tab is checked off
* AfterAttackRoll, which fires after an attack is rolled and the result is printed to chat

Added new conditions
* PowerContainsData, which lets you check the data of a power that's used in the OnPowerUse event
* AttackResultCondition, which lets you check if an attack roll is a hit, miss, critical hit, or fumble

Added new actions
* SendChatMessage, which prints a message to chat

Added compatibility for D&D 4e. Added a new action for 4e only that adds hp (normal or temp)

### CODE REVIEW

File: campaign/4e/char_power.xml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- 
+  Please see the license.html file included with this distribution for 
+  attribution and copyright information.
+-->
+
+<root>
+    <windowclass name="char_power" merge="join" ruleset="4E">
```
Comment (MeAndUnique) [inline campaign/4e/char_power.xml]: Code folding already struggles enough with FG XML, best to give it some assistance with consistent indenting.

In fact I'll just have this single comment for a blanket spaces to tabs request.

File: campaign/4e/char_power.xml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		<script>
+            local fActivatePowerOriginal;
+            function onInit()
```
Comment (MeAndUnique) [inline campaign/4e/char_power.xml]: I'm honestly a bit torn on this. On the one hand I'd like the larger script block to be in a lua file so the IDE can behave a bit better with it. On the other that would then make this file like 5 lines long, which seems perhaps a bit silly.

File: scripts/manager_action_damage_4e_tmt.lua [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+--
+
+rApplyHpCombatantAction = nil;
+
+function onInit()
+    initializeActions();
+end
+
+function initializeActions()
+    -- TODO: Add more parameters to this action
```
Comment (MeAndUnique) [inline scripts/manager_action_damage_4e_tmt.lua]: In either the next or second update I will be adding support for using/defining 5E power actions directly in triggers. So linking to a heal action on a character sheet or making one directly in the trigger using the same UI (or *maybe* a cousin version as I did for KNK, depending on how much the ruleset's tight coupling gets in the way).

If it is easy to do for 4E, I might deprecate this in favor of that, though I have no real familiarity with the ruleset so I may not take that on.

File: scripts/manager_char_4e_tmt.lua [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                sType = "string"
+            }
+        }
+    }
+
+    TriggerManager.defineCondition(rPowerContainsDataCondition);
+end
+
+function powerContainsDataCondition(rTriggerData, rEventData)
+    -- If contain property is empty, return true
```
Comment (MeAndUnique) [inline scripts/manager_char_4e_tmt.lua]: If I'm understanding correctly, I don't think there's a real use case for an empty sPowerContains parameter, other than user error/oversight. In that event I think have the trigger not work would alert them to that more quickly?

File: scripts/manager_char_4e_tmt.lua [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                aDefinedValues = {
+                    "power_used_property_name",
+                    "power_used_property_keywords",
+                    "power_used_property_recharge",
+                    "power_used_property_action",
+                    "power_used_property_range",
+                }
+            },
+            {
+                sName = "sPowerContains",
```
Comment (MeAndUnique) [inline scripts/manager_char_4e_tmt.lua]: This is perfectly usable as is and the below ramblings of mad man should not be brought into current scope....

What if there was also a combobox that was used instead and dynamically populated for the properties with discrete value sets!

File: scripts/manager_char_4e_tmt.lua (hunk 4 above)
Comment (Gtaray) [reply scripts/manager_char_4e_tmt.lua]: I can change it to that. The idea was if someone leaves it empty I don't want it throwing errors, but I'm fine either way.

File: scripts/manager_char_4e_tmt.lua (hunk 5 above)
Comment (Gtaray) [reply scripts/manager_char_4e_tmt.lua]: At least for 4e, there's dozens of possibilities for keywords alone, about a dozen different actions, half a dozen ranges. That might be a bit much, if I'm understanding the ramblings correctly.

File: scripts/manager_action_damage_4e_tmt.lua (hunk 3 above)
Comment (Gtaray) [reply scripts/manager_action_damage_4e_tmt.lua]: Healing is pretty different in 4e, so I'd have to see how it's implemented, but after that I can port it back to 4e and replace this if it's necessary.

File: campaign/4e/char_power.xml (hunk 2 above)
Comment (Gtaray) [reply campaign/4e/char_power.xml]: I moved them into their own scripts. It's cleaner at least.

File: scripts/manager_char_4e_tmt.lua (hunk 4 above)
Comment (MeAndUnique) [reply scripts/manager_char_4e_tmt.lua]: I leave it entirely to your judgement.

File: scripts/manager_char_4e_tmt.lua (hunk 5 above)
Comment (MeAndUnique) [reply scripts/manager_char_4e_tmt.lua]: Maybe an autofill then if fancy moon talk is happening. The base idea is that if there is a "correct" bit of text I'd like to keep my hammy little fingers away from mistyping it. Dealing narcotic damage lends a rather different vibe to an encounter after all.

**Pull request merged**